### PR TITLE
Add segment groups

### LINKED ISSUE #6954: Segments Groups
Similar to tree groups, there could be hierarchical segment groups. Group actions (via context menu) could be:

- [x] move selected segment into this group
- [ ] batch actions for groups #7164
  - [x] load/remove/refresh meshes for everything in a group 
  - [x] download all meshes in a group
  - [x] set color for all segments in group
  - [x] hide/show meshes
  - [ ] hide segments (not implemented yet)
- [ ] multi-selection

### PR DESCRIPTION
**Pull request by frcroth**
### URL of deployed dev instance (used for testing):
- https://segmentgroups.webknossos.xyz

### Steps to test:
- create a new annotation
- switch to the segments tab
- create some segments by brushing with some ids or simply click on existing segments
- right click on the root node in the segments tab to create one or multiple groups
- drag segments into the groups
- delete groups
- use undo/redo
- save & reload to check that everything was saved correctly

### Todo
- Front end (@philippotto)
  - [X] create groups
  - [X] delete group (recursively)
  - [X] renaming of groups
  - [X] dnd for groups and segments
  - [x] undo/redo for group mutations
  - [x] bugs
    - [x] segments and groups can be moved higher than the root which makes them disappear
    - [x] groups are not shown when no segments exist at all
    - [x] undo might still be buggy. I fixed some bugs, but I still managed to break it at some point. I didn't find a clear reproduction, though.
    - [X] item is moved to wrong group sometimes
  - [x] clean up
- [X] doublecheck export/import via NML (backend does this, for frontend, there is only a skeleton-related download, so no need to do adapt this in my opinion)
- ~~Handle merging of groups~~
- Fix e2e test

### Issues:
- fixes #6954 

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [ ] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

### CODE REVIEW

File: frontend/javascripts/oxalis/model/sagas/update_actions.ts
Comment (philippotto) [inline frontend/javascripts/oxalis/model/sagas/update_actions.ts:117]: I think we can use the same pattern for the other update actions, too, but I didn't want to hijack this PR further..

File: frontend/javascripts/oxalis/model/sagas/undo_saga.ts
Comment (philippotto) [inline frontend/javascripts/oxalis/model/sagas/undo_saga.ts:248]:
```suggestion
```

**Comment by philippotto** [comment]
@daniel-wer Feel free to already have a look at the PR. Most of it should be ready. Unfortunately, undo/redo is still a bit buggy in combination with group mutation, but I didn't really find good repro steps, yet. Maybe you have better luck :sweat_smile:

**Comment by philippotto** [comment]
@frcroth Is the backend ready for review?

**Comment by philippotto** [comment]
> It's basically finished except for undo/redo which is still a bit buggy in combination with group mutation, but I didn't really find good repro steps (that don't involve clicking for 2 minutes), yet.

Okay, I just found a way which is basically: set up segments and groups, delete everything recursively, undo, delete group (without recursion). That triggers the bug. I added a test for this which is currently failing and hope to find the actual problem. So, feel free to wait until I've got the fix for that :)

**Comment by philippotto** [comment]
@daniel-wer I think, I fixed the last remaining bug now :) There was simply one code spot where the `prevSegmentGroups` were not set. So, in the end it was a simple fix.

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:933]: This comment seems applicable to line 984

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:973]:
```suggestion
```

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:990]:
```suggestion
```

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:1028]:
```suggestion
```

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx
Comment (daniel-wer) [inline frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:1115]: Left over todo

**Review by daniel-wer — COMMENTED** [review]
Frontend review.

Very cool! :partying_face:  If we ever decide to also rewrite the skeleton list with the antd tree we might be able to reuse more of the code and obtain a more consistent UX.
The general functionality worked very well, but I encountered a crash and some rough edges during testing:

- [ ] Changing the color of a segment leads to a crash
- [ ] Importing an NML with segment groups through the frontend import fails or behaves wrongly, because the segment groups are not distinguished from the tree groups. So either the segment groups are imported as tree groups or the import fails, because of duplicate IDs. See this example NML:
  [NML_with_segment_groups.zip](https://github.com/scalableminds/webknossos/files/11287884/NML_with_segment_groups.zip)
  I understand that the frontend download doesn't include the segment groups, because it explicitly says "Download selected trees", but maybe it should import the segment groups and segments correctly? Otherwise, the import modal should state that only trees are imported, I think.
- [ ] When renaming a segment and trying to select the text using the mouse, the segment is dragged instead
- [ ] For some reason every element in the segment view now seems to have "---" placeholder tooltip :thinking: Is this some kind of antd shenanigan?
image

File: frontend/javascripts/oxalis/api/api_loader.ts
Comment (philippotto) [reply frontend/javascripts/oxalis/api/api_loader.ts:65]: in the request and model module, I used `typeof $instance` and here I'm using the class directly. Should both be fine :) It was important to me that the tests do an `import type` so that mocking is not disturbed.

File: frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration.spec.ts
Comment (philippotto) [reply frontend/javascripts/test/sagas/volumetracing/volumetracing_saga_integration.spec.ts:928]: I had "delete root group contents (including children)" in mind for the "recursion" part, but your suggestion makes the test more thorough. So, I'm adapting it :)

**Comment by philippotto** [comment]
>  Changing the color of a segment leads to a crash

Oops, this is broken on master currently. I pushed a fix. If I merge this PR on Monday, it should be fine, I think.

**Comment by philippotto** [comment]
> > Changing the color of a segment leads to a crash

> Oops, this is broken on master currently. I pushed a fix. If I merge this PR on Monday, it should be fine, I think.

@frcroth just pointed out that the back-end review is also pending, so I extracted the fix into https://github.com/scalableminds/webknossos/pull/7000 to not create an unnecessary pressure.

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx
Comment (philippotto) [reply frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:788]: You are right! I added this now as it was relatively simple :)

File: frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx
Comment (philippotto) [reply frontend/javascripts/oxalis/view/right-border-tabs/segments_tab/segments_view.tsx:299]: I decided that the segments and groups should simply be sorted by id. I think this is fair (because it's stable) for now :)

**Comment by philippotto** [comment]
The PR should be ready now from my side :) Not sure whether we want to wait for @daniel-wer's final review or maybe it's enough if somebody else does a last round of monkey testing?

> Importing an NML with segment groups through the frontend import fails or behaves wrongly, because the segment groups are not distinguished from the tree groups. So either the segment groups are imported as tree groups or the import fails, because of duplicate IDs. See this example NML:
[NML_with_segment_groups.zip](https://github.com/scalableminds/webknossos/files/11287884/NML_with_segment_groups.zip)
I understand that the frontend download doesn't include the segment groups, because it explicitly says "Download selected trees", but maybe it should import the segment groups and segments correctly? Otherwise, the import modal should state that only trees are imported, I think.

The import of skeleton groups should not be confused by segment groups anymore. I started implementing the import of segment groups, but it turned out to be a rabbit hole. In the end, the main problem was that it's not clear into which volume layer the segments should be imported (the id can change). Also, semantics-wise it is weird, because no actual volume data is imported. So, I decided together with @fm3, that the import should stick to skeletons and for everything else, the backend import should be used.

**Review by daniel-wer — APPROVED** [review]
LGTM, well done! I didn't notice any issues during retesting :tada:

**Pull request merged**
